File: scripts/check-refs.py
#!/usr/bin/env python3
"""Validate harness references for the opencode adapter (dotfiles deploy).

Convention: repo root maps 1:1 to ~/.config/opencode, except
  adapters/opencode/opencode.json -> opencode.json
  adapters/opencode/tui.json       -> tui.json
So {file:./...} in opencode.json is resolved target-relative against repo root.

Checks (source mode, default):
  1. opencode.json is valid JSON; every {file:PATH} resolves to an existing file.
  2. adapter.json is valid JSON; every non-glob include path exists;
     every glob prefix (before **) exists as a directory.
  3. commands/sdd-*.md absolute ~/.config/opencode/... paths map to repo files.

Bundle mode (--root dist/opencode): resolves the bundle's opencode.json
{file:} refs and commands absolute refs against the bundle root, so every
preset ships self-contained. Exit non-zero on any failure.
"""

import argparse
import json
import re
import sys
from pathlib import Path

REPO = Path(__file__).resolve().parent.parent
FAILURES: list[str] = []


def fail(msg: str) -> None:
    FAILURES.append(msg)
    print(f"FAIL: {msg}")


def check_opencode_refs(root: Path, cfg_rel: Path) -> int:
    cfg_path = root / cfg_rel
    cfg = json.loads(cfg_path.read_text())
    blob = json.dumps(cfg)
    refs = sorted(set(re.findall(r"\{file:([^}]+)\}", blob)))
    print(f"{cfg_rel}: {len(refs)} unique {{file:}} refs")
    ok = 0
    for ref in refs:
        target = (root / ref.lstrip("./")).resolve()
        # Guard: ref must stay inside the root (no ../ escapes past it)
        try:
            target.relative_to(root.resolve())
        except ValueError:
            fail(f"{{file:{ref}}} escapes root {root}")
            continue
        if target.is_file():
            ok += 1
        else:
            try:
                shown = target.relative_to(root.resolve())
            except ValueError:
                shown = target
            fail(f"{{file:{ref}}} -> {shown} MISSING")
    print(f"  resolved {ok}/{len(refs)}")
    return len(refs)


def check_adapter() -> None:
    root = REPO
    adapter_path = root / "adapters" / "opencode" / "adapter.json"
    adapter = json.loads(adapter_path.read_text())
    # compose sources: "layer:path"
    for target, sources in adapter.get("compose", {}).items():
        for src in sources:
            _, _, path = src.partition(":")
            if not (root / path).exists():
                fail(f"compose {target}: '{src}' MISSING")
    n = 0
    for comp, spec in adapter.get("components", {}).items():
        for pat in spec.get("include", []):
            n += 1
            if "**" in pat:
                prefix = pat.split("**")[0].rstrip("/")
                if prefix and not (root / prefix).exists():
                    fail(f"component {comp}: glob '{pat}' prefix MISSING")
            elif "*" in pat:
                if not list(root.glob(pat)):
                    fail(f"component {comp}: glob '{pat}' matches nothing")
            else:
                if not (root / pat).exists():
                    fail(f"component {comp}: '{pat}' MISSING")
    print(f"adapter.json: {n} include patterns checked")


def check_commands(cmds_dir: Path, resolve_root: Path) -> None:
    cmds = sorted(cmds_dir.glob("sdd-*.md"))
    n = 0
    for cmd in cmds:
        for m in re.finditer(r"~\/\.config\/opencode\/(\S+?)(?=[\s\"'`]|$)", cmd.read_text()):
            rel = m.group(1).rstrip(".,:;")
            n += 1
            if not (resolve_root / rel).exists():
                fail(f"{cmd.name}: ~/.config/opencode/{rel} MISSING")
    print(f"commands: {n} absolute path refs checked in {len(cmds)} files")


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--root", default=None, help="bundle dir to validate (default: repo source)")
    args = ap.parse_args()
    if args.root:
        # Bundle mode: opencode.json must be self-contained (pruned at render).
        # Command bodies keep source-mode validation: their absolute refs
        # assume the full deployed layout, not a minimal slice.
        root = Path(args.root)
        check_opencode_refs(root, Path("opencode.json"))
    else:
        check_opencode_refs(REPO, Path("adapters/opencode/opencode.json"))
        check_adapter()
        check_commands(REPO / "commands", REPO)
    if FAILURES:
        print(f"\n{len(FAILURES)} FAILURE(S)")
        return 1
    print("\nAll references OK")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/check-refs.mjs
#!/usr/bin/env node
// Validate harness references for the opencode adapter (dotfiles deploy).
//
// Convention: repo root maps 1:1 to ~/.config/opencode, except
//   adapters/opencode/opencode.json -> opencode.json
//   adapters/opencode/cli.json      -> cli.json
// So {file:./...} in opencode.json is resolved target-relative against repo root.
//
// Checks (source mode, default):
//   1. opencode.json is valid JSON; every {file:PATH} resolves to an existing file.
//   2. adapter.json is valid JSON; every non-glob include path exists;
//      every glob prefix (before **) exists as a directory.
//   3. commands/sdd-*.md absolute ~/.config/opencode/... paths map to repo files.
//
// Bundle mode (--root dist/opencode): resolves the bundle's opencode.json
// {file:} refs against the bundle root, so every preset ships self-contained.
// Exit non-zero on any failure.
//
// Cross-platform: no shell, no Python. Node 18+ built-ins only.

import {
  existsSync,
  readdirSync,
  readFileSync,
  statSync,
} from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const REPO = resolve(dirname(fileURLToPath(import.meta.url)), "..");
const FAILURES = [];

// True when `child` lives at or below `parent` (no `..` escapes past parent).
// path.relative(parent, parent) === ""; escapes start with "../" or are absolute.
function isInside(parent, child) {
  const rel = relative(parent, child);
  return rel !== "" && !rel.startsWith("..") && !/^[\\/]/.test(rel);
}

// Glob match check: returns true if `pattern` (with `*` only; no `**`) matches
// at least one file under `root`. Mirrors pathlib.Path.glob semantics:
// `*` matches any chars except the path separator.
function globHasMatch(root, pattern) {
  const regexStr =
    "^" +
    pattern
      .replace(/[.+^${}()|[\]\\]/g, "\\$&")
      .replace(/\*/g, "[^/]*") +
    "$";
  const regex = new RegExp(regexStr);
  let found = false;
  const walk = (absDir, relDir) => {
    if (found) return;
    let entries;
    try {
      entries = readdirSync(absDir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const e of entries) {
      if (e.name === ".DS_Store") continue;
      const rel = relDir === "" ? e.name : `${relDir}/${e.name}`;
      if (e.isDirectory()) walk(join(absDir, e.name), rel);
      else if (regex.test(rel)) {
        found = true;
        return;
      }
    }
  };
  walk(root, "");
  return found;
}

function fail(msg) {
  FAILURES.push(msg);
  console.log(`FAIL: ${msg}`);
}

function checkOpencodeRefs(root, cfgRel) {
  const cfgPath = join(root, cfgRel);
  const cfg = JSON.parse(readFileSync(cfgPath, "utf8"));
  const blob = JSON.stringify(cfg);
  const refs = [
    ...new Set(
      [...blob.matchAll(/\{file:([^}]+)\}/g)].map((m) => m[1]),
    ),
  ].sort();
  console.log(`${cfgRel}: ${refs.length} unique {file:} refs`);
  const rootResolved = resolve(root);
  let ok = 0;
  for (const ref of refs) {
    // Python: ref.lstrip("./") strips any leading "." or "/" chars.
    const stripped = ref.replace(/^[\.\/]+/, "");
    const target = resolve(join(root, stripped));
    if (!isInside(rootResolved, target)) {
      fail(`{file:${ref}} escapes root ${root}`);
      continue;
    }
    let isFile = false;
    try {
      isFile = statSync(target).isFile();
    } catch {
      isFile = false;
    }
    if (isFile) {
      ok += 1;
    } else {
      let shown = target;
      try {
        const rel = relative(rootResolved, target);
        if (rel && !rel.startsWith("..") && !/^[\\/]/.test(rel)) shown = rel;
      } catch {
        /* keep shown = target */
      }
      fail(`{file:${ref}} -> ${shown} MISSING`);
    }
  }
  console.log(`  resolved ${ok}/${refs.length}`);
  return refs.length;
}

function checkAdapter() {
  const root = REPO;
  const adapterPath = join(root, "adapters", "opencode", "adapter.json");
  const adapter = JSON.parse(readFileSync(adapterPath, "utf8"));
  // compose sources: "layer:path"
  for (const [target, sources] of Object.entries(adapter.compose ?? {})) {
    for (const src of sources) {
      const path = src.partition(":")[2];
      if (!existsSync(join(root, path))) {
        fail(`compose ${target}: '${src}' MISSING`);
      }
    }
  }
  let n = 0;
  for (const [comp, spec] of Object.entries(adapter.components ?? {})) {
    for (const pat of spec.include ?? []) {
      n += 1;
      if (pat.includes("**")) {
        const prefix = pat.split("**")[0].replace(/\/+$/, "");
        if (prefix && !existsSync(join(root, prefix))) {
          fail(`component ${comp}: glob '${pat}' prefix MISSING`);
        }
      } else if (pat.includes("*")) {
        if (!globHasMatch(root, pat)) {
          fail(`component ${comp}: glob '${pat}' matches nothing`);
        }
      } else {
        if (!existsSync(join(root, pat))) {
          fail(`component ${comp}: '${pat}' MISSING`);
        }
      }
    }
  }
  console.log(`adapter.json: ${n} include patterns checked`);
}

function checkCommands(cmdsDir, resolveRoot) {
  // Python: sorted(cmds_dir.glob("sdd-*.md"))
  const all = readdirSync(cmdsDir);
  const cmds = all.filter((f) => /^sdd-.*\.md$/.test(f)).sort();
  let n = 0;
  for (const name of cmds) {
    const text = readFileSync(join(cmdsDir, name), "utf8");
    // Python regex: r"~\/\.config\/opencode\/(\S+?)(?=[\s\"'`]|$)"
    const re = /~\/\.config\/opencode\/(\S+?)(?=[\s"'`]|$)/g;
    for (const m of text.matchAll(re)) {
      const rel = m[1].replace(/[.,:;]+$/, "");
      n += 1;
      if (!existsSync(join(resolveRoot, rel))) {
        fail(`${name}: ~/.config/opencode/${rel} MISSING`);
      }
    }
  }
  console.log(
    `commands: ${n} absolute path refs checked in ${cmds.length} files`,
  );
}

function parseArgs(argv) {
  const out = { root: null };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--root") {
      out.root = argv[++i];
    } else if (a === "--help" || a === "-h") {
      out.help = true;
    }
  }
  return out;
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  if (args.help) {
    console.log(
      "Usage: node scripts/check-refs.mjs [--root <bundleDir>]\n" +
        "  No --root: source mode (validates live repo).\n" +
        "  --root <path>: bundle mode (validates <path>/opencode.json).",
    );
    return 0;
  }
  if (args.root) {
    // Bundle mode: opencode.json must be self-contained (pruned at render).
    // Command bodies keep source-mode validation: their absolute refs
    // assume the full deployed layout, not a minimal slice.
    const root = resolve(args.root);
    checkOpencodeRefs(root, "opencode.json");
  } else {
    checkOpencodeRefs(REPO, join("adapters", "opencode", "opencode.json"));
    checkAdapter();
    checkCommands(join(REPO, "commands"), REPO);
  }
  if (FAILURES.length > 0) {
    console.log(`\n${FAILURES.length} FAILURE(S)`);
    return 1;
  }
  console.log("\nAll references OK");
  return 0;
}

process.exit(main());